[BUG] MinariExperienceReplay fails with remote datasets

## Describe the bug

Minari.load_dataset line fails if the dataset is downloaded by TorchRL, thus saved in TorchRL's cache, instead of the default Minari path.

Thus, MinariExperienceReplay fails:
https://github.com/pytorch/rl/blob/74abd6c38c065a17aeaa4bd617167055e434fdad/torchrl/data/datasets/minari_data.py#L473

## To Reproduce

Simply launch:
```python
import torchrl

ds = MinariExperienceReplay("D4RL/door/human-v2", batch_size=32)
```

```bash
    def test_torch_minari_experience_replay():
        """
        Tests creation of the torchrl MinariExperienceReplay object.
    
        According to their documentation, torchrl uses the MinariExperienceReplay
        object as a way to download datasets from the remote Farama server.
        """
    
        batch_size = 32
>       dataset = MinariExperienceReplay("D4RL/door/human-v2", batch_size=batch_size)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

tests/integrations/test_torch_rl.py:18: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/opt/anaconda3/envs/minari/lib/python3.12/site-packages/torchrl/data/datasets/minari_data.py:212: in __init__
    storage = self._download_and_preproc()
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/opt/anaconda3/envs/minari/lib/python3.12/site-packages/torchrl/data/datasets/minari_data.py:473: in _download_and_preproc
    dataset = minari.load_dataset(self.dataset_id)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

dataset_id = 'D4RL/door/human-v2', download = False

    def load_dataset(dataset_id: str, download: bool = False):
        """Retrieve Minari dataset from local database.
    
        Args:
            dataset_id (str): name id of Minari dataset
            download (bool): if `True` download the dataset if it is not found locally. Default to `False`.
    
        Returns:
            MinariDataset
        """
        file_path = get_dataset_path(dataset_id)
        data_path = os.path.join(file_path, "data")
    
        if not os.path.exists(data_path):
            if not download:
>               raise FileNotFoundError(
                    f"Dataset {dataset_id} not found locally at {file_path}. Use download=True to download the dataset."
                )
E               FileNotFoundError: Dataset D4RL/door/human-v2 not found locally at /var/folders/cn/j52qkn917b9b_j65205xwcmr0000gn/T/tmpj16ck37_/D4RL/door/human-v2. Use download=True to download the dataset.

minari/storage/local.py:51: FileNotFoundError
```

## Expected behavior

Should load it from the cache, if this was a TorchRL's download.


## System info

```python
import torchrl, numpy, sys
print(torchrl.__version__, numpy.__version__, sys.version, sys.platform)
```

gives: 

```
0.13.2 1.26.4 3.12.12 | packaged by Anaconda, Inc. | (main, Oct 21 2025, 20:07:49) [Clang 20.1.8 ] darwin
```

## Reason and Possible fixes

Wrap it as it is done for minari.download in the same file

## Checklist

- [x] I have checked that there is no similar issue in the repo (**required**)
- [x] I have read the [documentation](https://github.com/pytorch/rl/tree/main/docs/) (**required**)
- [x] I have provided a minimal working example to reproduce the bug (**required**)

--- a/test/libs/test_datasets.py
+++ b/test/libs/test_datasets.py
@@ -6,6 +6,7 @@ from __future__ import annotations
 
 import importlib.util
 import os
+import shutil
 import time
 import warnings
 from contextlib import nullcontext
@@ -726,6 +727,65 @@ class TestMinari:
         finally:
             if MINARI_DATASETS_PATH:
                 os.environ["MINARI_DATASETS_PATH"] = MINARI_DATASETS_PATH
+
+    @pytest.mark.skipif(
+        not _has_minari or not _has_gymnasium, reason="Minari or Gym not available"
+    )
+    def test_download_load_dataset_path(self, tmp_path, monkeypatch):
+        """Regression test for the ``MINARI_DATASETS_PATH`` handling on download.
+
+        ``_download_and_preproc`` downloads the dataset into a temporary directory
+        and then calls ``minari.load_dataset`` to read its metadata. Minari
+        resolves dataset locations through ``MINARI_DATASETS_PATH``; if that
+        variable is restored to a pre-existing value before the ``load_dataset``
+        call, Minari looks in the wrong directory and raises ``FileNotFoundError``.
+
+        Here we simulate a remote download by copying a locally-built dataset into
+        whatever directory torchrl points Minari at, while ``MINARI_DATASETS_PATH``
+        is set to an unrelated, empty cache. This used to fail before the fix.
+        """
+        import gymnasium
+        import minari
+        from minari import DataCollector
+
+        dataset_id = "cartpole/test-download-load-v1"
+
+        source_dir = tmp_path / "source"
+        monkeypatch.setenv("MINARI_DATASETS_PATH", str(source_dir))
+        env = DataCollector(gymnasium.make("CartPole-v1"), record_infos=True)
+        for _ in range(5):
+            env.reset(seed=123)
+            while True:
+                _, _, terminated, truncated, _ = env.step(env.action_space.sample())
+                if terminated or truncated:
+                    break
+        env.create_dataset(
+            dataset_id=dataset_id,
+            algorithm_name="RandomPolicy",
+            code_permalink="https://github.com/Farama-Foundation/Minari",
+            author="Farama",
+            author_email="contact@farama.org",
+            eval_env="CartPole-v1",
+        )
+
+        empty_cache = tmp_path / "empty_cache"
+        empty_cache.mkdir()
+        monkeypatch.setenv("MINARI_DATASETS_PATH", str(empty_cache))
+
+        def fake_download_dataset(dataset_id, *args, **kwargs):
+            dest = os.environ["MINARI_DATASETS_PATH"]
+            shutil.copytree(source_dir, dest, dirs_exist_ok=True)
+
+        monkeypatch.setattr(minari, "download_dataset", fake_download_dataset)
+
+        data = MinariExperienceReplay(
+            dataset_id=dataset_id,
+            batch_size=32,
+            root=str(tmp_path / "torchrl_root"),
+            download=True,
+        )
+        sample = data.sample()
+        assert sample.shape[0] == 32
 
     def test_correct_categorical_missions(self):
         try:

--- a/torchrl/data/datasets/minari_data.py
+++ b/torchrl/data/datasets/minari_data.py
@@ -10,8 +10,8 @@ import os.path
 import shutil
 import tempfile
 from collections import defaultdict
-from collections.abc import Callable
-from contextlib import nullcontext
+from collections.abc import Callable, Iterator
+from contextlib import contextmanager, nullcontext
 from dataclasses import asdict
 from pathlib import Path
 
@@ -38,6 +38,20 @@ from torchrl.envs.utils import _classproperty
 
 _has_tqdm = importlib.util.find_spec("tqdm", None) is not None
 _has_minari = importlib.util.find_spec("minari", None) is not None
+
+
+@contextmanager
+def _minari_datasets_path(path: str | Path) -> Iterator[None]:
+    prev = os.environ.get("MINARI_DATASETS_PATH")
+    os.environ["MINARI_DATASETS_PATH"] = str(path)
+    try:
+        yield
+    finally:
+        if prev is None:
+            os.environ.pop("MINARI_DATASETS_PATH", None)
+        else:
+            os.environ["MINARI_DATASETS_PATH"] = prev
+
 
 _NAME_MATCH = KeyDependentDefaultDict(lambda key: key)
 _NAME_MATCH["observations"] = "observation"
@@ -295,17 +309,8 @@ class MinariExperienceReplay(BaseDatasetExperienceReplay):
 
                 else:
                     # temporarily change the minari cache path
-                    prev_minari_datasets_path_save2 = os.environ.get(
-                        "MINARI_DATASETS_PATH"
-                    )
-                    os.environ["MINARI_DATASETS_PATH"] = tmpdir
-                    try:
+                    with _minari_datasets_path(tmpdir):
                         minari.download_dataset(dataset_id=self.dataset_id)
-                    finally:
-                        if prev_minari_datasets_path_save2 is not None:
-                            os.environ[
-                                "MINARI_DATASETS_PATH"
-                            ] = prev_minari_datasets_path_save2
 
                     parent_dir = Path(tmpdir) / self.dataset_id / "data"
 
@@ -469,7 +474,12 @@ class MinariExperienceReplay(BaseDatasetExperienceReplay):
                         from torchrl.collectors.utils import split_trajectories
 
                         td_data = split_trajectories(td_data).memmap_(self.data_path)
-                with open(self.metadata_path, "w") as metadata_file:
+                load_path = (
+                    prev_minari_datasets_path if self.load_from_local_minari else tmpdir
+                )
+                with open(
+                    self.metadata_path, "w"
+                ) as metadata_file, _minari_datasets_path(load_path):
                     dataset = minari.load_dataset(self.dataset_id)
                     self.metadata = asdict(dataset.spec)
                     self.metadata["observation_space"] = _spec_to_dict(

